zebra-state/src/service/finalized_state/disk_format/upgrade/add_subtrees.rs

```diff
@@ -180,76 +290,78 @@ pub fn run(
+        panic!(
+            "missing or bad subtree(s): sapling: {sapling_result:?}, orchard: {orchard_result:?}"
+        );
     }
 }
 
 /// Check that Sapling note commitment subtrees were correctly added.
 ///
-/// # Panics
-///
-/// If a note commitment subtree is missing or incorrect.
-fn check_sapling_subtrees(db: &ZebraDb) -> bool {
+/// Returns an error if a note commitment subtree is missing or incorrect.
+fn check_sapling_subtrees(db: &ZebraDb) -> Result<(), &'static str> {
     let Some(NoteCommitmentSubtreeIndex(mut first_incomplete_subtree_index)) =
         db.sapling_tree().subtree_index()
     else {
-        return true;
+        return Ok(());
     };
 
     // If there are no incomplete subtrees in the tree, also expect a subtree for the final index.
     if db.sapling_tree().is_complete_subtree() {
         first_incomplete_subtree_index += 1;
     }
 
-    let mut is_valid = true;
+    let mut result = Ok(());
     for index in 0..first_incomplete_subtree_index {
         // Check that there's a continuous range of subtrees from index [0, first_incomplete_subtree_index)
         let Some(subtree) = db.sapling_subtree_by_index(index) else {
-            error!(index, "missing subtree");
-            is_valid = false;
+            result = Err("missing subtree");
+            error!(?result, index);
             continue;
         };
 
         // Check that there was a sapling note at the subtree's end height.
         let Some(tree) = db.sapling_tree_by_height(&subtree.end) else {
-            error!(?subtree.end, "missing note commitment tree at subtree completion height");
-            is_valid = false;
+            result = Err("missing note commitment tree at subtree completion height");
+            error!(?result, ?subtree.end);
             continue;
         };
 
         // Check the index and root if the sapling note commitment tree at this height is a complete subtree.
         if let Some((index, node)) = tree.completed_subtree_index_and_root() {
             if subtree.index != index {
-                error!("completed subtree indexes should match");
-                is_valid = false;
+                result = Err("completed subtree indexes should match");
+                error!(?result);
             }
 
             if subtree.node != node {
-                error!("completed subtree roots should match");
-                is_valid = false;
+                result = Err("completed subtree roots should match");
+                error!(?result);
             }
         }
         // Check that the final note has a greater subtree index if it didn't complete a subtree.
         else {
             let Some(prev_tree) = db.sapling_tree_by_height(&subtree.end.previous()) else {
-                error!(?subtree.end, "missing note commitment tree at subtree completion height");
-                is_valid = false;
+                result = Err("missing note commitment tree at subtree completion height");
```

Comment:
###### Optional Suggestion

```suggestion
                result = Err("missing note commitment tree below subtree completion height");
```